fix: log to console using structured logging

minherz opened the pull request:
Replace package `console-log-level` with `@google-cloud/logging-min`. Use LogSync class of the logging package to print structured logging to `stdout` to capture the correct severity level.
Also handle a case of printing multi-line log that is captured as multiple log entries.

minherz committed:
fix: use nodejs-logging

Replaces 'console-log-level' with '@google-cloud/logging-min' to write logs to stdout.

minherz committed:
chore: refactor processing log messages

Rolls back to accept a string message for Logger methods.
Replaces new line with its escape notation '\\n'.
Refactors single location that prints array to print a single string.

minherz committed:
chore: add initialization required for logging

minherz committed:
chore: update node versioning

minherz committed:
chore: align node versioning with one in nodejs-logging

minherz committed:
chore: align engine version with nodejs-logging

minherz commented:
@bcoe can you help me to troubleshoot the reasons for check failures, please?

minherz committed:
Merge branch 'main' into minherz/b/182594320

minherz commented:
Test fails will be fixed with https://github.com/mapbox/node-pre-gyp/issues/623

losalex commented on src/logger.ts:
```diff
@@ -12,36 +12,69 @@
+  }
+
+  private toOneLine(msg: string): string {
+    return msg.replace('\n', '\\n');
```
Do you need to handle '\r' as well?

minherz replied:
I do not think that standalone `\r` should be handled. Would the following handle Windows new lines IYO:
```java
var temp = msg.replace('\r\n', '\\r\\n');
return temp.replace('\n', '\\n');
```

minherz replied:
I saw it a common practice of capturing multiline payload in Json is to replace with an anchor that can be used to pretty print the text later.
It is hard to restore the original multi-line string later (e.g. callstacks). Such work will require from a reader a lot of effort and manual work. With the anchor, a simple find-and-replace in an editor or with a script/code will do the work.
Would removing \r means that in certain OS (e.g. Windows) restoring will not work?

minherz replied:
Just updated the code. Please, see the new version that handles new lines in both Window and Linux formats.

losalex commented on src/logger.ts:
```diff
@@ -12,36 +12,69 @@
+    }
+  }
+
+  private toOneLine(msg: string): string {
```
Is one line a requirement?

minherz replied:
Yes. Printing to STDOUT is captured one line per log entry.

losalex commented on src/logger.ts:
```diff
@@ -12,36 +12,69 @@
+    if (level === undefined) {
+      level = defaultConfig.logLevel;
+    }
+    if (level < 0) {
```
I guess you should define a constants for this

minherz replied:
Can you be more specific? What values should be set to constants?

minherz replied:
Added constants and a comment that explains where it comes from.

losalex commented on src/logger.ts:
```diff
@@ -12,36 +12,69 @@
+    if (level === undefined) {
+      level = defaultConfig.logLevel;
+    }
+    if (level < 0) {
```
Is this intentionally? I wonder in case of "invalid" value should you use `defaultConfig.logLevel`?

minherz replied:
I am not sure what you mean here. If the level is undefined then the default level is used. The conditions for level value beyond the range are handled exactly the same way they were handled before the change. The change does not modify existing interfaces or behavior, it only replaces the engine that prints logs to STDOUT behind it.

minherz replied:
It would change the current behavior.

losalex reviewed (approved):
Some comments inside

losalex reviewed (approved):
Left some comments, see if you agree with those

minherz committed:
chore: fix minimal node 10 version constraint

ensure to pass test to not accept node@10.4.0

minherz committed:
chore: code refactoring

add new line at the end of the package.json.
add constants for max and min logging levels.
add handling windows new line in the multiline log payload.

minherz committed:
Merge branch 'main' into minherz/b/182594320
